### PR TITLE
[Merged by Bors] - feat(RingTheory/AlgebraicIndependent): add `AlgebraicIndependent.aevalEquivField` and other results

### PR DESCRIPTION
**Pull request by acmepjz**
This PR mainly adds

- `AlgebraicIndependent.aevalEquivField`: canonical isomorphism between rational function field and the intermediate field generated by algebraically independent elements.
- `AlgebraicIndependent.reprField`: canonical map from the intermediate field generated by an algebraic independent family into the rational function field.

Other results:

- `AlgebraicIndependent.aeval_of_algebraicIndependent`: if `x = {x_i : A | i : ι}` and `f = {f_i : MvPolynomial ι R | i : ι}` are algebraically independent over `R`, then `{f_i(x) | i : ι}` is also algebraically independent over `R`.
- `AlgebraicIndependent.of_aeval`: conversely, if `{f_i(x) | i : ι}` is algebraically independent over `R`, then `{f_i : MvPolynomial ι R | i : ι}` is also algebraically independent over `R`.
- `AlgebraicIndependent.polynomial_aeval_of_transcendental`: if `{x_i : A | i : ι}` is algebraically independent over `R`, and for each `i`, `f_i : R[X]` is transcendental over `R`, then `{f_i(x_i) | i : ι}` is also algebraically independent over `R`.
- `AlgebraicIndependent.isEmpty_of_isAlgebraic`: if `A/R` is algebraic, then all algebraically independent families are empty.
- `IsTranscendenceBasis.isEmpty_iff_isAlgebraic`: if `x` is a transcendence basis of `A/R`, then it is empty if and only if `A/R` is algebraic.
- `IsTranscendenceBasis.nonempty_iff_transcendental`: if `x` is a transcendence basis of `A/R`, then it is not empty if and only if `A/R` is transcendental.
- `IsTranscendenceBasis.isAlgebraic_field`: if `x` is a transcendence basis of `E / F`, then the field extension `E / F(x)` is algebraic.

Also add `[of_]ringHom_of_comp_eq` for `[Algebra.]Transcendental` and `AlgebraicIndependent`.

Also add some results for `Subalgebra` inspired by `IntermediateField.isAlgebraic_iff`. Some of them are not back-ported to `IntermediateField` yet; may be for another PR if they are useful.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

- [x] depends on: #18661
- [x] depends on: #18984

### CODE REVIEW

File: Mathlib/RingTheory/AlgebraicIndependent.lean
Comment (alreadydone) [inline Mathlib/RingTheory/AlgebraicIndependent.lean:720]:
```suggestion
    (RingHom.id A) ?_ Function.injective_id (by ext; rfl)
  rintro ⟨_, r, rfl⟩; exact ⟨r, rfl⟩
```
and I think we should extract the lemma `Algebra.IsAlgebraic (↥⊥) A ↔ Algebra.IsAlgebraic R A`.

**Review by alreadydone — APPROVED** [review]

File: Mathlib/RingTheory/Localization/FieldRange.lean
Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FieldRange.lean]: Done.

File: Mathlib/RingTheory/Localization/FieldRange.lean
Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FieldRange.lean:31]: Do you mean a `IsScalarTower` version?

File: Mathlib/RingTheory/AlgebraicIndependent.lean
Comment (acmepjz) [reply Mathlib/RingTheory/AlgebraicIndependent.lean]: Done. The results are extracted into 'Mathlib/RingTheory/Algebraic.lean`. Note that some of the corresponding results for `IntermediateField` are not added yet; they should probably go to `FieldTheory/Adjoin.lean`.

File: Mathlib/RingTheory/Algebraic.lean
Comment (alreadydone) [inline Mathlib/RingTheory/Algebraic.lean:402]:
```suggestion
  halg.of_ringHom_of_comp_eq (algebraMap R (⊥ : Subalgebra R S))
    (RingHom.id S) (by rintro ⟨_, r, rfl⟩; exact ⟨r, rfl⟩) Function.injective_id rfl

theorem isAlgebraic_bot_iff (h : Function.Injective (algebraMap R S)) {x : S} :
    _root_.IsAlgebraic (⊥ : Subalgebra R S) x ↔ _root_.IsAlgebraic R x :=
  isAlgebraic_ringHom_iff_of_comp_eq (Algebra.botEquivOfInjective h).symm (RingHom.id S)
    Function.injective_id (by rfl)
```

File: Mathlib/RingTheory/Algebraic.lean
Comment (alreadydone) [inline Mathlib/RingTheory/Algebraic.lean:419]:
```suggestion
instance algebra_isAlgebraic_bot [Nontrivial R] : Algebra.IsAlgebraic R (⊥ : Subalgebra R S) :=
  ⟨by rintro ⟨_, x, rfl⟩; exact isAlgebraic_algebraMap _⟩
```

**Review by alreadydone — APPROVED** [review]
LGTM modulo #18984. The import increase is mainly due to AlgebraicIndependent importing FieldRange importing FieldTheory.Adjoin, but I think we're gonna import FieldTheory.Adjoin into AlgebraicIndependent sooner or later. Notice that AlgebraicIndependent is currently almost a leaf file, being transitively imported by only 2 other files.

**Comment by acmepjz** [comment]
> I think we're gonna import FieldTheory.Adjoin into AlgebraicIndependent sooner or later.

I think so, as most part of algebraic independent theory is about fields.

Maybe later we need to split AlgebraicIndependent into several files: subalgebra part, field part, and transcendence basis & transcendence degree part.

> Notice that AlgebraicIndependent is currently almost a leaf file, being transitively imported by only 2 other files.

Yeah, but once transcendence degree is defined, it should be used extensively in algebraic geometry. We also need separable transcendence basis & separably generated, etc. etc.

**Review by alreadydone — COMMENTED** [review]
Thanks :tada: 
maintainer merge

**Review by alreadydone — APPROVED** [review]

**Pull request closed without merging**
